components/canvas/webgl_thread.rs

```diff
@@ -0,0 +1,1191 @@
+        match msg {
+            WebGLMsg::CreateContext(size, attributes, result_sender) => {
+                let result = self.create_webgl_context(size, attributes);
+                result_sender.send(result.map(|(id, limits, share_mode)|
+                    WebGLCreateContextResult {
+                        sender: WebGLMsgSender::new(id, webgl_chan.clone()),
+                        limits: limits,
+                        share_mode: share_mode,
+                    }
+                )).unwrap();
+            },
+            WebGLMsg::ResizeContext(ctx_id, size, sender) => {
+                self.resize_webgl_context(ctx_id, size, sender);
+            },
+            WebGLMsg::RemoveContext(ctx_id) => {
+                self.remove_webgl_context(ctx_id);
+            },
+            WebGLMsg::WebGLCommand(ctx_id, command) => {
+                self.handle_webgl_command(ctx_id, command);
+            },
+            WebGLMsg::WebVRCommand(ctx_id, command) => {
+                self.handle_webvr_command(ctx_id, command);
+            },
+            WebGLMsg::Lock(ctx_id, sender) => {
+                self.handle_lock(ctx_id, sender);
+            },
+            WebGLMsg::Unlock(ctx_id) => {
+                self.handle_unlock(ctx_id);
+            },
+            WebGLMsg::UpdateWebRenderImage(ctx_id, sender) => {
+                self.handle_update_wr_image(ctx_id, sender);
+            },
+            WebGLMsg::Exit => {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Handles a WebGLCommand for a specific WebGLContext
+    fn handle_webgl_command(&mut self, context_id: WebGLContextId, command: WebGLCommand) {
+        let ctx = &self.contexts[&context_id];
+        if Some(context_id) != self.current_bound_webgl_context_id {
+            ctx.make_current();
+            self.current_bound_webgl_context_id = Some(context_id);
+        }
+        ctx.apply_command(command);
+    }
+
+    /// Handles a WebVRCommand for a specific WebGLContext
+    fn handle_webvr_command(&mut self, context_id: WebGLContextId, command: WebVRCommand) {
+        if Some(context_id) != self.current_bound_webgl_context_id {
+            self.contexts[&context_id].make_current();
+            self.current_bound_webgl_context_id = Some(context_id);
+        }
+
+        let texture = match command {
+            WebVRCommand::SubmitFrame(..) => {
+                self.cached_context_info.get(&context_id)
+            },
+            _ => None
+        };
+        self.webvr_compositor.as_mut().unwrap().handle(command, texture.map(|t| (t.texture_id, t.size)));
+    }
+
+    /// Handles a lock external callback received from webrender::ExternalImageHandler
+    fn handle_lock(&mut self, context_id: WebGLContextId, sender: WebGLSender<(u32, Size2D<i32>)>) {
+        let ctx = &self.contexts[&context_id];
+        if Some(context_id) != self.current_bound_webgl_context_id {
+            ctx.make_current();
```

Comment:
(It may return the current context, if the extra hashmap lookup worries you, I guess).